zerver/models.py

```diff
@@ -3307,12 +3312,40 @@ def to_dict(self) -> Dict[str, Any]:
 post_delete.connect(flush_used_upload_space_cache, sender=Attachment)
 
 
-def validate_attachment_request(user_profile: UserProfile, path_id: str) -> Optional[bool]:
+def validate_attachment_request_for_spectator_access(
+    realm: Realm, attachment: Attachment
+) -> Optional[bool]:
+    if attachment.realm != realm:
+        return False
+
+    if settings.RATE_LIMITING:
+        try:
+            from zerver.lib.rate_limiter import rate_limit_spectator_attachment_access_by_file
+
+            rate_limit_spectator_attachment_access_by_file(attachment.path_id)
```

Comment:
Why check this before checking `attachment.is_web_public`?  This is just going to take up keyspace in the rate limiting store even for attachments which are not currently public.

Reply:
It probably doesn't matter, in that it's not easy to guess the URL of a valid attachment in the first place to trigger this check, but it does seem easier to reason about with the other order.

Reply:
Changed this ordering.